mlflow/store/artifact/s3_artifact_repo.py

```diff
@@ -188,13 +189,48 @@ def _verify_listed_object_contains_artifact_path_prefix(listed_object_path, arti
                     artifact_path=artifact_path, object_path=listed_object_path
                 )
             )
-
+     
     def _download_file(self, remote_file_path, local_path):
         (bucket, s3_root_path) = data.parse_s3_uri(self.artifact_uri)
         s3_full_path = posixpath.join(s3_root_path, remote_file_path)
         s3_client = self._get_s3_client()
+
+        # Check if the file already exists
+        if os.path.exists(local_path):
+            # Etag comparison from: https://stackoverflow.com/questions/1775816/how-to-get-the-md5sum-of-a-file-on-amazons-s3
+            obj_dict = s3_client.get_object(Bucket=bucket, Key=s3_full_path)
+            etag = (obj_dict['ETag'])
+            # if the file with the same hash exists do not download it again
+            if self._etag_compare(local_path, etag):
+                return
         s3_client.download_file(bucket, s3_full_path, local_path)
 
+    @staticmethod
+    def _md5_checksum(filename):
+        m = hashlib.md5()
+        with open(filename, 'rb') as f:
+            for data in iter(lambda: f.read(1024 * 1024), b''):
+                m.update(data)
+        return m.hexdigest()
+
+    @staticmethod
+    def _etag_checksum(filename, chunk_size=8 * 1024 * 1024):
+        md5s = []
+        with open(filename, 'rb') as f:
+            for data in iter(lambda: f.read(chunk_size), b''):
+                md5s.append(hashlib.md5(data).digest())
+        m = hashlib.md5(b"".join(md5s))
+        return '{}-{}'.format(m.hexdigest(), len(md5s))
```

Comment:
`len(md5s)` seems to depend on `chunk_size`, but we can't know the actual chunk size that was used to upload the file. Accordingly, it seems like this checksum won't be valid for multi-part uploads.